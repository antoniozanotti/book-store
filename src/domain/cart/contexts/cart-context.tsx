"use client";

import {
  createContext,
  useContext,
  useReducer,
  ReactNode,
  useEffect,
} from "react";
import { CartType } from "../types/cart-type";
import { CartContextType } from "../types/cart-context-type";
import { useLocalStorage } from "usehooks-ts";
import { useProductsByIdsQuery } from "@/domain/product/queries/use-products-by-ids-query";
import { cartReducer } from "../reducers/cart-reducer";

const initialState: CartType = { items: [] };

const CartContext = createContext<CartContextType>({
  state: initialState,
  dispatch: () => {},
  getCartItemsQty: () => {},
  cartItemsSubtotal: () => {},
});

export function CartProvider({ children }: { children: ReactNode }) {
  const [localState, setLocalState] = useLocalStorage<CartType>(
    "cart",
    initialState
  );
  const [state, dispatch] = useReducer(cartReducer, localState);
  const cartItems = state.items;
  const responses = useProductsByIdsQuery(cartItems.map((item) => item.id));

  function getCartItemsQty(): number {
    const cartItems = state.items;
    return cartItems.reduce((acc, cur) => {
      return acc + cur.quantity;
    }, 0);
  }

  function cartItemsSubtotal(): number {
    let subTotal = 0;
    cartItems.forEach((item) => {
      responses.some((response) => {
        const product = response.data;
        if (!product) {
          return true;
        }
        if (item.id == product.id) {
          subTotal += parseInt(product.price.toString()) * item.quantity;
          return true;
        }
      });
    });
    return subTotal;
  }

  useEffect(() => {
    setLocalState(state);
  }, [state, setLocalState]);

  return (
    <CartContext.Provider
      value={{ state, dispatch, getCartItemsQty, cartItemsSubtotal }}
    >
      {children}
    </CartContext.Provider>
  );
}

export function useCartContext(): CartContextType {
  return useContext(CartContext);
}
