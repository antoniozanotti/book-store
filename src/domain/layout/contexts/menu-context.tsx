"use client";

import { createContext, useContext, useReducer, ReactNode } from "react";
import { MenuType } from "../types/menu-type";
import { MenuContextType } from "../types/menu-context-type";
import { menuReducer } from "../reducers/menu-reducer";

const initialState: MenuType = { isOpen: false };

const MenuContext = createContext<MenuContextType>({
  state: initialState,
  dispatch: () => {},
});

export function MenuProvider({ children }: { children: ReactNode }) {
  const [state, dispatch] = useReducer(menuReducer, initialState);

  return (
    <MenuContext.Provider value={{ state, dispatch }}>
      {children}
    </MenuContext.Provider>
  );
}

export function useMenuContext(): MenuContextType {
  return useContext(MenuContext);
}
