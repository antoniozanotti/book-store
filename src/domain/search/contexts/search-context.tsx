"use client";

import { createContext, useContext, useReducer, ReactNode } from "react";
import { SearchContextType } from "../types/search-context-type";
import { useSearchParams } from "next/navigation";
import { getSearchParams } from "../services/search-service";
import { searchReducer } from "../reducers/search-reducer";

const SearchContext = createContext<SearchContextType>({} as SearchContextType);

export function SearchProvider({ children }: { children: ReactNode }) {
  const searchParams = useSearchParams();
  const { keyword, categories, minPrice, maxPrice, minYear, maxYear, orderBy } =
    getSearchParams(searchParams);
  const initialStateBySearchParams = {
    keyword: keyword,
    categories: categories,
    minprice: minPrice,
    maxprice: maxPrice,
    minyear: minYear,
    maxyear: maxYear,
    orderby: orderBy,
  };
  const [state, dispatch] = useReducer(
    searchReducer,
    initialStateBySearchParams
  );

  return (
    <SearchContext.Provider value={{ state, dispatch }}>
      {children}
    </SearchContext.Provider>
  );
}

export function useSearchContext(): SearchContextType {
  return useContext(SearchContext);
}
