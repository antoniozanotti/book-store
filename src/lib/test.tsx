import React, { ReactElement, ReactNode } from "react";
import {
  render,
  RenderOptions,
  RenderResult,
  Screen,
} from "@testing-library/react";

function AllTheProviders({ children }: { children: React.ReactNode }) {
  return <div data-testid="container">{children}</div>;
}

function customRender(
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
): RenderResult {
  return render(ui, { wrapper: AllTheProviders, ...options });
}

function firstComponent(screen: Screen) {
  return screen.getByTestId("container").firstChild;
}

export * from "@testing-library/react";
export * from "@testing-library/user-event";
export { customRender as render, firstComponent };
import "@testing-library/jest-dom";
import { usePathname } from "next/navigation";

jest.mock("next/navigation", () => {
  return {
    useRouter() {
      return { push: () => {} };
    },
    useParams: () => ({}),
    notFound: () => null,
    usePathname: () => null,
    useSearchParams: () => null,
  };
});
