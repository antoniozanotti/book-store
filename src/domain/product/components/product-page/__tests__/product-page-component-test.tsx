import { render, screen } from "@/lib/test";
import { ProductPageComponent } from "../product-page-component";
import { ProductPageProvider } from "@/domain/product/contexts/product-page-context";
import { productWithBookAndCategoriesMock } from "@/domain/product/__tests__/product-with-book-and-categories-mock";
import { Layout } from "lucide-react";
import { LayoutProvider } from "@/domain/layout/contexts/layout-context";
import { categoriesWithPageMock } from "@/domain/category/__tests__/categories-with-page-mock";

describe("ProductPageComponent", () => {
  it("should render", async () => {
    const { container } = render(
      <LayoutProvider categories={categoriesWithPageMock}>
        <ProductPageProvider product={productWithBookAndCategoriesMock}>
          <ProductPageComponent />
        </ProductPageProvider>
      </LayoutProvider>
    );
    expect(container).toMatchSnapshot();
  });
});
