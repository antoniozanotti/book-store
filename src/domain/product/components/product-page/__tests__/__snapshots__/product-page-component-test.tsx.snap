// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductPageComponent should render 1`] = `
<div>
  <div
    data-testid="container"
  >
    <div
      class="flex flex-row"
    >
      <aside
        class="basis-1/4 hidden lg:block"
      >
        <nav
          aria-label="Categories"
          class="p-10"
        >
          <h2
            class="font-primary text-pretty text-2xl"
          >
            Categories
          </h2>
          <ul
            class="text-sm flex flex-col divide-y"
          >
            <li
              class="py-2 "
            >
              <a
                href="/category/arts-and-entertainment"
                title="Arts and Entertainment"
              >
                Arts and Entertainment
              </a>
            </li>
            <li
              class="py-2 "
            >
              <a
                href="/category/business-and-technology"
                title="Business and Technology"
              >
                Business and Technology
              </a>
            </li>
            <li
              class="py-2 "
            >
              <a
                href="/category/graphic-novels"
                title="Graphic Novels"
              >
                Graphic Novels
              </a>
            </li>
            <li
              class="py-2 "
            >
              <a
                href="/category/history-biography-and-social-science"
                title="History, Biography, and Social Science"
              >
                History, Biography, and Social Science
              </a>
            </li>
            <li
              class="py-2 "
            >
              <a
                href="/category/home-and-lifestyle"
                title="Home and Lifestyle"
              >
                Home and Lifestyle
              </a>
            </li>
            <li
              class="py-2 "
            >
              <a
                href="/category/horror"
                title="Horror"
              >
                Horror
              </a>
            </li>
            <li
              class="py-2 "
            >
              <a
                href="/category/languages-and-travel"
                title="Languages and Travel"
              >
                Languages and Travel
              </a>
            </li>
            <li
              class="py-2 "
            >
              <a
                href="/category/lgbtqia"
                title="LGBTQIA+"
              >
                LGBTQIA+
              </a>
            </li>
            <li
              class="py-2 "
            >
              <a
                href="/category/literature-and-poetry"
                title="Literature and Poetry"
              >
                Literature and Poetry
              </a>
            </li>
            <li
              class="py-2 "
            >
              <a
                href="/category/mystery"
                title="Mystery"
              >
                Mystery
              </a>
            </li>
            <li
              class="py-2 "
            >
              <a
                href="/category/romance"
                title="Romance"
              >
                Romance
              </a>
            </li>
            <li
              class="py-2 "
            >
              <a
                href="/category/science-and-math"
                title="Science and Math"
              >
                Science and Math
              </a>
            </li>
            <li
              class="py-2 "
            >
              <a
                href="/category/science-fiction-and-fantasy"
                title="Science Fiction and Fantasy"
              >
                Science Fiction and Fantasy
              </a>
            </li>
            <li
              class="py-2 "
            >
              <a
                href="/category/spirituality-and-philosophy"
                title="Spirituality and Philosophy"
              >
                Spirituality and Philosophy
              </a>
            </li>
            <li
              class="py-2 "
            >
              <a
                href="/category/wellness-sports-and-outdoors"
                title="Wellness, Sports, and Outdoors"
              >
                Wellness, Sports, and Outdoors
              </a>
            </li>
            <li
              class="py-2 "
            >
              <a
                href="/category/young-adult"
                title="Young Adult"
              >
                Young Adult
              </a>
            </li>
          </ul>
        </nav>
      </aside>
      <main
        class="basis-3/4 grow p-10"
      >
        <div
          class="lg:flex"
        >
          <div
            class="lg:basis-2/3"
          >
            <div>
              <h1
                class="font-primary text-pretty text-4xl sm:text-5xl md:text-6xl lg:text-7xl"
              >
                The Four Agreements: A Practical Guide to Personal Freedom
              </h1>
              <address>
                by 
                <a
                  class="underline"
                  href="/author/don-miguel-ruiz"
                  rel="author"
                  title="Don Miguel Ruiz"
                >
                  Don Miguel Ruiz
                </a>
              </address>
              <figure
                class="bg-green-300 w-full sm:w-[374px] h-[497px] relative overflow-hidden mt-4"
              >
                <img
                  alt="The Four Agreements: A Practical Guide to Personal Freedom"
                  class="transition-all object-cover w-full h-full"
                  data-nimg="1"
                  decoding="async"
                  height="497"
                  loading="lazy"
                  sizes="374px"
                  src="https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_3840,h_5102,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0"
                  srcset="https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_16,h_21,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 16w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_32,h_42,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 32w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_48,h_63,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 48w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_64,h_85,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 64w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_96,h_127,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 96w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_128,h_170,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 128w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_256,h_340,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 256w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_384,h_510,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 384w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_640,h_850,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 640w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_750,h_996,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 750w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_828,h_1100,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 828w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_1080,h_1435,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 1080w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_1200,h_1594,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 1200w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_1920,h_2551,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 1920w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_2048,h_2721,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 2048w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_3840,h_5102,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 3840w"
                  style="color: transparent;"
                  width="374"
                />
              </figure>
              <h2
                class="font-primary text-pretty text-3xl mt-10"
              >
                Synopses
              </h2>
              <p>
                Canto eum versus somniculosus mollitia conforto celebrer iste. Surculus facere vix acceptus. Cohaero demo sufficio venio autus animi ex adversus quam.
              </p>
            </div>
            <div>
              <h2
                class="font-primary text-pretty text-3xl mt-10"
              >
                Product Details
              </h2>
              <table
                class="text-left"
              >
                <tbody>
                  <tr>
                    <th
                      scope="row"
                    >
                      Year
                    </th>
                    <td>
                      2024
                    </td>
                  </tr>
                  <tr>
                    <th
                      scope="row"
                    >
                      Pages
                    </th>
                    <td>
                      400
                    </td>
                  </tr>
                  <tr>
                    <th
                      scope="row"
                    >
                      Publisher
                    </th>
                    <td>
                      Hoeger - Swaniawski
                    </td>
                  </tr>
                  <tr>
                    <th
                      scope="row"
                    >
                      ISBN
                    </th>
                    <td>
                      978-1-315-92574-5
                    </td>
                  </tr>
                  <tr>
                    <th
                      scope="row"
                    >
                      Category
                    </th>
                    <td>
                      <a
                        class="underline"
                        href="/category/spirituality-and-philosophy"
                        title="Spirituality and Philosophy"
                      >
                        Spirituality and Philosophy
                      </a>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
          <div
            class="mt-10 lg:mt-0 lg:basis-1/3"
          >
            <div
              class="flex flex-col text-center border p-4"
              data-testid="add-to-cart-component"
            >
              <span
                class="text-2xl p-2"
              >
                $302.00
              </span>
              <button
                class="bg-primary-900 font-bold p-4 text-primary-100 p-2"
              >
                ADD TO CART
              </button>
              <span
                class="bg-primary-100 p-2 text-sm text-primary-900"
              >
                Ships in 1 to 3 days
              </span>
            </div>
          </div>
        </div>
        <div
          class="mt-4"
        >
          <h3
            class="font-primary text-pretty text-3xl mt-10"
          >
            Spirituality and Philosophy
             Books You May Also Like
          </h3>
          <div
            class="mt-2"
          >
            <div
              class="flex flex-wrap justify-center gap-4"
            >
              <article
                class="p-5 w-full sm:w-[227px] h-[400px] text-sm text-pretty"
              >
                <a
                  href="/product/the-prophet"
                  title="The Prophet by Kahlil Gibran"
                >
                  <figure
                    class="bg-green-300 w-full sm:w-[187px] h-[249px] relative overflow-hidden"
                  >
                    <img
                      alt="The Prophet"
                      class="transition-all hover:-mt-[20px] w-full h-[269px] object-cover"
                      data-nimg="1"
                      decoding="async"
                      height="269"
                      loading="lazy"
                      sizes="640px"
                      src="https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_3840,h_5523,g_auto/f_auto/q_auto/v1/product/the-prophet?_a=BAVARSDW0"
                      srcset="https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_16,h_23,g_auto/f_auto/q_auto/v1/product/the-prophet?_a=BAVARSDW0 16w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_32,h_46,g_auto/f_auto/q_auto/v1/product/the-prophet?_a=BAVARSDW0 32w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_48,h_69,g_auto/f_auto/q_auto/v1/product/the-prophet?_a=BAVARSDW0 48w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_64,h_92,g_auto/f_auto/q_auto/v1/product/the-prophet?_a=BAVARSDW0 64w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_96,h_138,g_auto/f_auto/q_auto/v1/product/the-prophet?_a=BAVARSDW0 96w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_128,h_184,g_auto/f_auto/q_auto/v1/product/the-prophet?_a=BAVARSDW0 128w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_256,h_368,g_auto/f_auto/q_auto/v1/product/the-prophet?_a=BAVARSDW0 256w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_384,h_552,g_auto/f_auto/q_auto/v1/product/the-prophet?_a=BAVARSDW0 384w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_640,h_920,g_auto/f_auto/q_auto/v1/product/the-prophet?_a=BAVARSDW0 640w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_750,h_1078,g_auto/f_auto/q_auto/v1/product/the-prophet?_a=BAVARSDW0 750w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_828,h_1191,g_auto/f_auto/q_auto/v1/product/the-prophet?_a=BAVARSDW0 828w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_1080,h_1553,g_auto/f_auto/q_auto/v1/product/the-prophet?_a=BAVARSDW0 1080w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_1200,h_1726,g_auto/f_auto/q_auto/v1/product/the-prophet?_a=BAVARSDW0 1200w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_1920,h_2761,g_auto/f_auto/q_auto/v1/product/the-prophet?_a=BAVARSDW0 1920w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_2048,h_2946,g_auto/f_auto/q_auto/v1/product/the-prophet?_a=BAVARSDW0 2048w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_3840,h_5523,g_auto/f_auto/q_auto/v1/product/the-prophet?_a=BAVARSDW0 3840w"
                      style="color: transparent;"
                      width="187"
                    />
                  </figure>
                </a>
                <p
                  class="font-primary text-pretty text-2xl line-clamp-2 mt-2"
                >
                  <a
                    href="/product/the-prophet"
                    title="The Prophet by Kahlil Gibran"
                  >
                    The Prophet
                  </a>
                </p>
                <address
                  class="line-clamp-2"
                >
                  by 
                  Kahlil Gibran
                </address>
              </article>
              <article
                class="p-5 w-full sm:w-[227px] h-[400px] text-sm text-pretty"
              >
                <a
                  href="/product/tao-of-pooh"
                  title="Tao of Pooh by Benjamin Hoff"
                >
                  <figure
                    class="bg-green-300 w-full sm:w-[187px] h-[249px] relative overflow-hidden"
                  >
                    <img
                      alt="Tao of Pooh"
                      class="transition-all hover:-mt-[20px] w-full h-[269px] object-cover"
                      data-nimg="1"
                      decoding="async"
                      height="269"
                      loading="lazy"
                      sizes="640px"
                      src="https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_3840,h_5523,g_auto/f_auto/q_auto/v1/product/tao-of-pooh?_a=BAVARSDW0"
                      srcset="https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_16,h_23,g_auto/f_auto/q_auto/v1/product/tao-of-pooh?_a=BAVARSDW0 16w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_32,h_46,g_auto/f_auto/q_auto/v1/product/tao-of-pooh?_a=BAVARSDW0 32w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_48,h_69,g_auto/f_auto/q_auto/v1/product/tao-of-pooh?_a=BAVARSDW0 48w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_64,h_92,g_auto/f_auto/q_auto/v1/product/tao-of-pooh?_a=BAVARSDW0 64w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_96,h_138,g_auto/f_auto/q_auto/v1/product/tao-of-pooh?_a=BAVARSDW0 96w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_128,h_184,g_auto/f_auto/q_auto/v1/product/tao-of-pooh?_a=BAVARSDW0 128w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_256,h_368,g_auto/f_auto/q_auto/v1/product/tao-of-pooh?_a=BAVARSDW0 256w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_384,h_552,g_auto/f_auto/q_auto/v1/product/tao-of-pooh?_a=BAVARSDW0 384w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_640,h_920,g_auto/f_auto/q_auto/v1/product/tao-of-pooh?_a=BAVARSDW0 640w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_750,h_1078,g_auto/f_auto/q_auto/v1/product/tao-of-pooh?_a=BAVARSDW0 750w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_828,h_1191,g_auto/f_auto/q_auto/v1/product/tao-of-pooh?_a=BAVARSDW0 828w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_1080,h_1553,g_auto/f_auto/q_auto/v1/product/tao-of-pooh?_a=BAVARSDW0 1080w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_1200,h_1726,g_auto/f_auto/q_auto/v1/product/tao-of-pooh?_a=BAVARSDW0 1200w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_1920,h_2761,g_auto/f_auto/q_auto/v1/product/tao-of-pooh?_a=BAVARSDW0 1920w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_2048,h_2946,g_auto/f_auto/q_auto/v1/product/tao-of-pooh?_a=BAVARSDW0 2048w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_3840,h_5523,g_auto/f_auto/q_auto/v1/product/tao-of-pooh?_a=BAVARSDW0 3840w"
                      style="color: transparent;"
                      width="187"
                    />
                  </figure>
                </a>
                <p
                  class="font-primary text-pretty text-2xl line-clamp-2 mt-2"
                >
                  <a
                    href="/product/tao-of-pooh"
                    title="Tao of Pooh by Benjamin Hoff"
                  >
                    Tao of Pooh
                  </a>
                </p>
                <address
                  class="line-clamp-2"
                >
                  by 
                  Benjamin Hoff
                </address>
              </article>
              <article
                class="p-5 w-full sm:w-[227px] h-[400px] text-sm text-pretty"
              >
                <a
                  href="/product/the-road-less-traveled-a-new-psychology-of-love-traditional-values-and-spiritual-growth"
                  title="The Road Less Traveled: A New Psychology of Love, Traditional Values, and Spiritual Growth by M. Scott Peck"
                >
                  <figure
                    class="bg-green-300 w-full sm:w-[187px] h-[249px] relative overflow-hidden"
                  >
                    <img
                      alt="The Road Less Traveled: A New Psychology of Love, Traditional Values, and Spiritual Growth"
                      class="transition-all hover:-mt-[20px] w-full h-[269px] object-cover"
                      data-nimg="1"
                      decoding="async"
                      height="269"
                      loading="lazy"
                      sizes="640px"
                      src="https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_3840,h_5523,g_auto/f_auto/q_auto/v1/product/the-road-less-traveled-a-new-psychology-of-love-traditional-values-and-spiritual-growth?_a=BAVARSDW0"
                      srcset="https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_16,h_23,g_auto/f_auto/q_auto/v1/product/the-road-less-traveled-a-new-psychology-of-love-traditional-values-and-spiritual-growth?_a=BAVARSDW0 16w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_32,h_46,g_auto/f_auto/q_auto/v1/product/the-road-less-traveled-a-new-psychology-of-love-traditional-values-and-spiritual-growth?_a=BAVARSDW0 32w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_48,h_69,g_auto/f_auto/q_auto/v1/product/the-road-less-traveled-a-new-psychology-of-love-traditional-values-and-spiritual-growth?_a=BAVARSDW0 48w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_64,h_92,g_auto/f_auto/q_auto/v1/product/the-road-less-traveled-a-new-psychology-of-love-traditional-values-and-spiritual-growth?_a=BAVARSDW0 64w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_96,h_138,g_auto/f_auto/q_auto/v1/product/the-road-less-traveled-a-new-psychology-of-love-traditional-values-and-spiritual-growth?_a=BAVARSDW0 96w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_128,h_184,g_auto/f_auto/q_auto/v1/product/the-road-less-traveled-a-new-psychology-of-love-traditional-values-and-spiritual-growth?_a=BAVARSDW0 128w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_256,h_368,g_auto/f_auto/q_auto/v1/product/the-road-less-traveled-a-new-psychology-of-love-traditional-values-and-spiritual-growth?_a=BAVARSDW0 256w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_384,h_552,g_auto/f_auto/q_auto/v1/product/the-road-less-traveled-a-new-psychology-of-love-traditional-values-and-spiritual-growth?_a=BAVARSDW0 384w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_640,h_920,g_auto/f_auto/q_auto/v1/product/the-road-less-traveled-a-new-psychology-of-love-traditional-values-and-spiritual-growth?_a=BAVARSDW0 640w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_750,h_1078,g_auto/f_auto/q_auto/v1/product/the-road-less-traveled-a-new-psychology-of-love-traditional-values-and-spiritual-growth?_a=BAVARSDW0 750w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_828,h_1191,g_auto/f_auto/q_auto/v1/product/the-road-less-traveled-a-new-psychology-of-love-traditional-values-and-spiritual-growth?_a=BAVARSDW0 828w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_1080,h_1553,g_auto/f_auto/q_auto/v1/product/the-road-less-traveled-a-new-psychology-of-love-traditional-values-and-spiritual-growth?_a=BAVARSDW0 1080w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_1200,h_1726,g_auto/f_auto/q_auto/v1/product/the-road-less-traveled-a-new-psychology-of-love-traditional-values-and-spiritual-growth?_a=BAVARSDW0 1200w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_1920,h_2761,g_auto/f_auto/q_auto/v1/product/the-road-less-traveled-a-new-psychology-of-love-traditional-values-and-spiritual-growth?_a=BAVARSDW0 1920w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_2048,h_2946,g_auto/f_auto/q_auto/v1/product/the-road-less-traveled-a-new-psychology-of-love-traditional-values-and-spiritual-growth?_a=BAVARSDW0 2048w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_3840,h_5523,g_auto/f_auto/q_auto/v1/product/the-road-less-traveled-a-new-psychology-of-love-traditional-values-and-spiritual-growth?_a=BAVARSDW0 3840w"
                      style="color: transparent;"
                      width="187"
                    />
                  </figure>
                </a>
                <p
                  class="font-primary text-pretty text-2xl line-clamp-2 mt-2"
                >
                  <a
                    href="/product/the-road-less-traveled-a-new-psychology-of-love-traditional-values-and-spiritual-growth"
                    title="The Road Less Traveled: A New Psychology of Love, Traditional Values, and Spiritual Growth by M. Scott Peck"
                  >
                    The Road Less Traveled: A New Psychology of Love, Traditional Values, and Spiritual Growth
                  </a>
                </p>
                <address
                  class="line-clamp-2"
                >
                  by 
                  M. Scott Peck
                </address>
              </article>
              <article
                class="p-5 w-full sm:w-[227px] h-[400px] text-sm text-pretty"
              >
                <a
                  class="flex items-center h-full"
                  href="/category/spirituality-and-philosophy"
                  title="Spirituality and Philosophy"
                >
                  <span>
                    See all books from 
                    Spirituality and Philosophy
                    .
                  </span>
                </a>
              </article>
            </div>
          </div>
        </div>
      </main>
    </div>
  </div>
</div>
`;
